drivers/mpu9250/MPU9250.cpp

```diff
@@ -383,130 +548,71 @@ void MPU9250::_measure()
 	// Use 1 MHz for normal registers.
 	_setBusFrequency(SPI_FREQUENCY_1MHZ);
 	uint8_t int_status = 0;
+
 	int result = _readReg(MPUREG_INT_STATUS, int_status);
 
 	if (result != 0) {
-		m_synchronize.lock();
-		++m_sensor_data.error_counter;
-		m_synchronize.unlock();
+		_counters.errors++;
 		return;
 	}
 
 	if (int_status & BITS_INT_STATUS_FIFO_OVERFLOW) {
 		reset_fifo();
 
-		m_synchronize.lock();
-		++m_sensor_data.fifo_overflow_counter;
+		_counters.fifo_overflows++;
 		DF_LOG_ERR("FIFO overflow");
-		m_synchronize.unlock();
 
 		return;
 	}
 
-	int size_of_fifo_packet;
-
-	if (_mag_enabled) {
-		size_of_fifo_packet = sizeof(fifo_packet_with_mag);
-
-	} else {
-		size_of_fifo_packet = sizeof(fifo_packet);
-	}
-
 	// Get FIFO byte count to read and floor it to the report size.
-	int bytes_to_read = get_fifo_count() / size_of_fifo_packet
-			    * size_of_fifo_packet;
-
-	// It looks like the FIFO doesn't actually deliver at 8kHz like it is supposed to.
-	// Therefore, we need to adapt the interval which we pass on to the integrator.
-	// The filtering is to lower the jitter that could result through the calculation
-	// because of the fact that the bytes we fetch per _measure() cycle varies.
-	_packets_per_cycle_filtered = (0.95f * _packets_per_cycle_filtered) + (0.05f * (bytes_to_read / size_of_fifo_packet));
-
-	if (bytes_to_read < 0) {
-		m_synchronize.lock();
-		++m_sensor_data.error_counter;
-		m_synchronize.unlock();
-		return;
-	}
+	int bytes_to_read;
+	do {
+		bytes_to_read = get_fifo_count();
+		if (bytes_to_read < 0) {
+			_counters.errors++;
+			DF_LOG_ERR("bytes_to_read: %d\n", bytes_to_read);
+			return;
+		} else if (bytes_to_read < _size_of_fifo_packet) {
+			/* TODO: replace usleep with nanosleep, check for interrupts, limit the sleep time */
+			usleep(1);
```

Comment:
I doubt we should sleep in here, see comment above. These changes complicate everything a lot in my opinion.